### ISSUE TITLE
'Should' does not log arrays with $null elements accurately

### ISSUE THREAD

**Issue by johndog** [pester/Pester#1448]
## 1. General summary of the issue

Run this command:

` echo a b c $null | should be "a","b","c"   

Result:
```
Expected @('a', 'b', 'c'), but got @('a', 'b', 'c').
At C:\Program Files\WindowsPowerShell\Modules\Pester\4.9.0\Functions\Assertions\Should.ps1:169 char:9
+         throw ( New-ShouldErrorRecord -Message $testResult.FailureMes ...
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidResult: (System.Collections.Hashtable:Hashtable) [], Exception
    + FullyQualifiedErrorId : PesterAssertionFailed

```

Specifically, the first line logs that the arrays are the same when they aren't (or the test would have failed).

## 2. Describe Your Environment

Pester version     : 4.9.0 C:\Program Files\WindowsPowerShell\Modules\Pester\4.9.0\Pester.psm1
PowerShell version : 5.1.19555.1001
OS version         : Microsoft Windows NT 10.0.19555.0

## 3. Expected Behavior

If the test fails, the line:

`Expected @('a', 'b', 'c'), but got @('a', 'b', 'c').`

should indicate why, instead of printing two equal arrays and saying they aren't equal.

Note that this test passes:

`echo a b c $null | should be "a","b","c",$null  `

And this test actually outputs \<empty\>:

```
> echo a b c $null | should be "a","b","c",'' 

Expected @('a', 'b', 'c', <empty>), but got @('a', 'b', 'c').
```

## 4. Possible Solution
Suggested outcomes:

1) `Expected @('a', 'b', 'c'), but got @('a', 'b', 'c', $null).`
2) `Expected @('a', 'b', 'c'), but got @('a', 'b', 'c', '').`

## 6. Context
This makes it difficult to debug a test in which an array evaluates to an empty element.

**Comment by nohwnd** [comment]
This looks like an issue with Format-Nicely, but looking at the code it should work correctly. Would you be willing to PR a fix for this? 

https://github.com/pester/Pester/blob/master/Dependencies/Format/Format.psm1#L11 <- this code should call Format-Null which would print `$null`.

I would start by adding tests here https://github.com/pester/Pester/blob/master/Dependencies/Format/Format.Tests.ps1#L21
And here ttps://github.com/pester/Pester/blob/master/Dependencies/Format/Format.Tests.ps1#L137

if you can't figure it out, please send a `WIP:` PR so I can take over :)

**Comment by johndog** [comment]
No time to submit a PR tonight, but the problem is in the Select:

```
> 1,2,$null,3 | select -first 5
1
2
3
> (1,2,$null,3 | select -first 5).Count
3
```

So before it gets to Format-Nicely, Select-Object removes the $null (Format-Collection, Format.psm1 line 11).

I moved the select to the end of the pipeline and now it works fine locally:

```
    '@(' + (($Value | % { Format-Nicely -Value $_ -Pretty:$Pretty } | Select -First $Limit ) -join $separator ) + <# $(if ($trimmed) {' +' + [string]($count-$limit)}) + #> ')'
```

**Comment by johndog** [comment]
BTW, I also gathered evidence for my suspicion that aborting at the end of the pipeline is perf-wise equivalent:

```
PS > 1..100 | % {write-host $_; $_} | select -first 5
1
1
2
2
3
3
4
4
5
5
```

If the whole pipeline didn't abort, I'd expect the write-host to keep outputting up to 100.

**Comment by nohwnd** [comment]
Okay I see, there should also be an elipsis `...` or something that indicates that the collection was trimmed. I see the code there commented out, not sure why I did that, it probably looked weird / confusing with the +10, so the fixed version should see how many items there are and if only 2 are remaining over the limit we should write them anyway, otherwise print `+2 other` or `...2 more`, or just `...`. Agreed? Select should be replaced with foreach imo.

**Comment by fflaten** [comment]
Still broken in 5.3.0-alpha1

**Comment by nohwnd** [comment]
This is a good first issue to tackle, you just need to add tests for this in the Format tests, and implement the fix for the $null (should show up as `$null` in the output), and add ellipsis if the array was trimmed (if that is no done yet).

**Comment by ArmaanMcleod** [comment]
Hi @nohwnd, @fflaten and @johndog, I had some time to work on this and submitted a PR. I apologise If this is already taken by anyone, should have asked here first 😅 .

**Comment by nohwnd** [comment]
Thanks! Merged!

**Comment by ArmaanMcleod** [comment]
@nohwnd Can probably close this issue. Some reason I didn't link my PR and it didn't close automatically 😄

**Comment by fflaten** [comment]
Closing. Fixed by #1961 and available in latest alpha-version of 5.3.0. Currently presents $null like suggestion 2 from OP.
Thanks @ArmaanMcleod !
